staging/src/k8s.io/component-helpers/scheduling/corev1/helpers.go

```diff
@@ -30,3 +36,111 @@ func PodPriority(pod *v1.Pod) int32 {
 	// name of the pod was empty. So, we resolve to the static default priority.
 	return 0
 }
+
+// NodeSelectorRequirementsAsSelector converts the []NodeSelectorRequirement API type into a labels.Selector.
+func nodeSelectorRequirementsAsSelector(nsr []v1.NodeSelectorRequirement) (labels.Selector, error) {
+	if len(nsr) == 0 {
+		return labels.Nothing(), nil
+	}
+	selector := labels.NewSelector()
+	for _, expr := range nsr {
+		var op selection.Operator
+		switch expr.Operator {
+		case v1.NodeSelectorOpIn:
+			op = selection.In
+		case v1.NodeSelectorOpNotIn:
+			op = selection.NotIn
+		case v1.NodeSelectorOpExists:
+			op = selection.Exists
+		case v1.NodeSelectorOpDoesNotExist:
+			op = selection.DoesNotExist
+		case v1.NodeSelectorOpGt:
+			op = selection.GreaterThan
+		case v1.NodeSelectorOpLt:
+			op = selection.LessThan
+		default:
+			return nil, fmt.Errorf("%q is not a valid node selector operator", expr.Operator)
+		}
+		r, err := labels.NewRequirement(expr.Key, op, expr.Values)
+		if err != nil {
+			return nil, err
+		}
+		selector = selector.Add(*r)
+	}
+	return selector, nil
+}
+
+// MatchNodeSelectorTerms checks whether the node labels or name match node selector terms in ORed;
+// nil or empty term matches no objects.
+func MatchNodeSelectorTerms(
+	node *v1.Node,
+	nodeSelectorTerms []v1.NodeSelectorTerm,
+) (bool, error) {
+	errors := make([]error, 0)
+	for _, req := range nodeSelectorTerms {
+		// nil or empty term selects no objects
+		if len(req.MatchExpressions) == 0 && len(req.MatchFields) == 0 {
+			continue
+		}
+
+		if len(req.MatchExpressions) != 0 {
+			labelSelector, err := nodeSelectorRequirementsAsSelector(req.MatchExpressions)
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+			if labelSelector == nil || !labelSelector.Matches(labels.Set(node.Labels)) {
+				continue
+			}
+		}
+
+		if len(req.MatchFields) != 0 {
+			fieldSelector, err := nodeSelectorRequirementsAsFieldSelector(req.MatchFields)
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+			if fieldSelector == nil || !fieldSelector.Matches(fields.Set{
+				"metadata.name": node.Name,
+			}) {
+				continue
+			}
+		}
+
+		return true, apierrors.NewAggregate(errors)
```

Comment:
returning `true, non-nil error` here is confusing and likely to break callers that first check error before checking the true/false return value... if the contract of the method is that terms are OR-ed, and we get a true here, one of the terms matched, so this should not be an error, right?